modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/util/PageHandler.java

```diff
@@ -0,0 +1,478 @@
+
+        try {
+            if ((pageAddr = readLock(pageMem, groupId, pageId, page, lsnr)) == 0L) {
+                return lockFailed;
+            }
+
+            PageIo io = pageMem.ioRegistry().resolve(pageAddr);
+
+            return h.run(groupId, pageId, page, pageAddr, io, arg, intArg, statHolder);
+        } finally {
+            if (pageAddr != 0L) {
+                readUnlock(pageMem, groupId, pageId, page, pageAddr, lsnr);
+            }
+        }
+    }
+
+    /**
+     * Acquires the read lock on the page.
+     *
+     * @param pageMem Page memory.
+     * @param groupId Group ID.
+     * @param pageId  Page ID.
+     * @param page    Page pointer.
+     * @param lsnr    Lock listener.
+     * @return Page address or {@code 0} if acquiring failed.
+     */
+    public static long readLock(
+            PageMemory pageMem,
+            int groupId,
+            long pageId,
+            long page,
+            PageLockListener lsnr
+    ) {
+        lsnr.onBeforeReadLock(groupId, pageId, page);
+
+        long pageAddr = pageMem.readLock(groupId, pageId, page);
+
+        lsnr.onReadLock(groupId, pageId, page, pageAddr);
+
+        return pageAddr;
+    }
+
+    /**
+     * Releases acquired read lock.
+     *
+     * @param pageMem  Page memory.
+     * @param groupId  Group ID.
+     * @param pageId   Page ID.
+     * @param page     Page pointer.
+     * @param pageAddr Page address.
+     * @param lsnr     Lock listener.
+     */
+    public static void readUnlock(
+            PageMemory pageMem,
+            int groupId,
+            long pageId,
+            long page,
+            long pageAddr,
+            PageLockListener lsnr
+    ) {
+        lsnr.onReadUnlock(groupId, pageId, page, pageAddr);
+
+        pageMem.readUnlock(groupId, pageId, page);
+    }
+
+    /**
+     * Initializes a new page.
+     *
+     * @param pageMem    Page memory.
+     * @param groupId    Group ID.
+     * @param pageId     Page ID.
+     * @param init       IO for new page initialization.
+     * @param lsnr       Lock listener.
+     * @param statHolder Statistics holder to track IO operations.
+     * @throws IgniteInternalCheckedException If failed.
+     * @see PageIo#initNewPage(long, long, int)
+     */
+    public static void initPage(
+            PageMemory pageMem,
+            int groupId,
+            long pageId,
+            PageIo init,
+            PageLockListener lsnr,
+            IoStatisticsHolder statHolder
+    ) throws IgniteInternalCheckedException {
+        Boolean res = writePage(
+                pageMem,
+                groupId,
+                pageId,
+                lsnr,
+                PageHandler.NO_OP,
+                init,
+                null,
+                0,
+                FALSE,
+                statHolder
+        );
+
+        assert res != FALSE;
+    }
+
+    /**
+     * Executes handler under the read lock or returns {@code lockFailed} if lock failed.
```

Comment:
In the code, `writeLock()` is taken, should it be 'write lock' in the javadoc?

Reply:
oopsie